docs/src/architecture/08_concepts/cardano_chain_follower/follower_instances.md

```diff
@@ -0,0 +1,122 @@
+
+## Implementation details
+
+Application modules can subscribe to events from Tip, Genesis OR a particular block within the block chain.
+There are a number of considerations which should be handled to make stream selection efficient.
+
+There is also a case where a stream is requested to "continue" and it needs special handling to be efficient.
+
+### Maximizing the number of listeners to an event stream
+
+We want to maximize the number of listeners to individual event streams.
+This will improve efficiency overall.
+
+If two applications both request a stream of blocks from mainnet, from Genesis.
+Then ideally they both listen to the same stream.
+However, both applications can start asynchronously,
+and there is no guarantee when each application or module will subscribe to the stream.
+
+This can cause the following situation (These event all happen in rapid succession in this example):
+
+1. App 1 subscribes to genesis.
+2. Connection to Node occurs, and block 1 is received, and sent to App 1.
+3. App 2 subscribes to genesis.
+4. The first connection can not be used because we are already past genesis on that stream, so we need to create a second stream.
+
+The problem here is we have two streams, but they are only 1 block apart.  
+Further, based on timing it may be that either exceeds the other depending on how fast their individual connection to the relay is.
+This is not ideal.
+
+There are a number of ways this can be solved.
+
+1. Use a block cache for recent blocks, so that subscriptions that are close in time can use the same stream by
+   re-using blocks from the cache.
+2. Delay when we connect a stream to allow for the maximum number of subscribers.
+3. Optimizing the streams on the fly.
+    * For example if 2 streams are at different points, but they could catch up and become equal.
+    the moment they become equal one of the streams is closed and all block data is now taken from a single stream.
+    * This is likely to occur, for example a follower from Tip and Genesis.
+    Once the Genesis follower has reached tip we now have two followers from Tip when only 1 is required.
+4. An ideal solution may need multiple techniques.
+
+*Note: There may be other techniques and these are just some obvious solutions to the issue.*
+
+For the initial implementation of the Follower, we will employ technique 2 and a limited version of Technique 3 ONLY.
+
+#### Delaying starting of the streams
+
+When a stream is subscribed, and it results in a new stream, then the connection to that relay node will not happen for 5 seconds.
+Any further subscriptions to the same stream can then be added to the pending connection.
+Once the connection is made, then further subscriptions will only be added to it if they are valid for that connection.
+Otherwise a second connection is established, again with its own 5 second start delay.
+
+This delay occurs regardless of where the stream is to start from Tip, Genesis or a particular block.
```

Comment:
What if instead of just waiting 5 seconds we kept track of where each follower is at and on a subscription request we:

1. assigned the closest one that is at a point previous to the point the subscriber wants to follow from
2. Waited for 5 seconds for it to reach the point the subscriber wants
3. If it didn't reach that point in 5 seconds, we spawn another follower at that point

Does that make sense?

Reply:
That only works if your asfter somethin that's in the future.  If I request Genesis, and the follower immediately gets block 1 and 2 and sends those events, its now a follower pointing to block 3.
If another module asks to follow from genesis our existing follower is already past genesis so we now need another follower that's running 3 blocks behind.  The delay is to ensure we don't get that scenario, which is more likely.

For example, I have a module that says "Give me the blocks from genesis", and then a different module says "Give me the transactions from genesis".  We could end up with two followers, when what we really wanted was one.

Reply:
there are other ways, but a 5 second delay is an easy one to ensure followers requested at about the same time are synced.  Your suggestion would also be viable if I asked for genesis in one follower, and block 10 in another, we could park the block 10 one until the genesis follower catches up.  But I think this is a less likely scenario, though worth documenting.

What I am not trying to achieve here is optimal follower setup, but one that catches the most likely events and then we can see how it behaves and iterate.  I don't want to over think/over engineer it too much up front.

In fact in a first PR I would 100% accept no optimization of the followers at all, just so we get the logic right in its simplest form.  Then adding progressive optimizations to the running followers could be follow up.  Also for M2 we don;t need these optimizations.